Added Template BigCapital

Ahson-Shaikh opened the pull request:
BigCapital is a powerful, self-hosted platform designed for managing financial
  assets and operations. It provides a clean and intuitive interface, supports
  plugins for extended functionality, and offers detailed financial tracking
  analytics. With BigCapital, you can maintain control over your investments and
  branding. The application is accessible via your host on ports 8080 and 8081.
  This container is based on Docker Baseimage KasmVNC, providing additional
  environment variables and configurations for enhanced flexibility. It also
  supports the docker manifest for multi-platform awareness, ensuring the
  correct image for your architecture.

deiucanta commented on templates/bigcapital/index.ts:
```diff
@@ -0,0 +1,165 @@
+      ].join("\n"),
+      source: {
+        type: "image",
+        image: "dockeriddonuts/bigcapital-migration:1.1",
```
this is still inline, you need to extract it to meta.yaml

also, please provide all the custom Dockerfiles as a comment in the PR

Ahson-Shaikh replied:
@deiucanta Fixed

Ahson-Shaikh commented:
Custom Dockerfiles used in this: 
dockeriddonuts/bigcapital-nginx:1.0
dockeriddonuts/bigcapital-migration:1.1
dockeriddonuts/bigcapital-mongo:1.0
dockeriddonuts/bigcapital-mysql:1.0
Official Images: 
bigcapitalhq/webapp:v0.22.0
bigcapitalhq/server:v0.22.0

deiucanta commented:
@Ahson-Shaikh I need to see the content of each Dockerfile

Ahson-Shaikh commented:
This is the Nginx-Dockerfile

`FROM nginx:1.25-alpine


RUN apk add --no-cache gettext


COPY nginx.template.conf /etc/nginx/templates/
COPY docker-entrypoint.sh /


RUN sed -i 's/\r$//' /docker-entrypoint.sh && \
    chmod +x /docker-entrypoint.sh

ENTRYPOINT ["/docker-entrypoint.sh"]`

This is the nginx.template.conf

`user  nginx;
worker_processes  auto;

events {
    worker_connections  1024;
}

http {
   
    upstream dynamic_server {
        server ${DYNAMIC_SERVER}:${DYNAMIC_PORT};  
        keepalive 32;
    }

   
    upstream webapp {
        server ${WEBAPP_SERVER}:${WEBAPP_PORT};  
        keepalive 32;
    }

    server {
        listen 80;
        server_name localhost;

        
        location /api {
            proxy_pass http://dynamic_server;
            proxy_http_version 1.1;
            proxy_set_header Connection "";
            
            
            proxy_connect_timeout 250ms;
            proxy_send_timeout 250ms;
            proxy_read_timeout 250ms;
        }

        location / {
            proxy_pass http://webapp;
            proxy_http_version 1.1;
            proxy_set_header Connection "";
            
            proxy_connect_timeout 250ms;
            proxy_send_timeout 250ms;
            proxy_read_timeout 250ms;
        }
    }`
    
   This is entry-point.sh
   
`#!/bin/sh

envsubst < /etc/nginx/templates/nginx.template.conf > /etc/nginx/nginx.conf

exec nginx -g 'daemon off;'`

Ahson-Shaikh commented:
Big-Capital Migration

I replaced the variable in CMD directive for checking for DB host.

```
FROM bigcapitalhq/server:latest as build

ARG DB_HOST= \
  DB_USER= \
  DB_PASSWORD= \
  DB_CHARSET= \
  # System database.
  SYSTEM_DB_NAME= \
  SYSTEM_DB_PASSWORD= \
  SYSTEM_DB_USER= \
  SYSTEM_DB_HOST= \
  SYSTEM_DB_CHARSET=

ENV DB_HOST=$DB_HOST \
  DB_USER=$DB_USER \
  DB_PASSWORD=$DB_PASSWORD \
  DB_CHARSET=$DB_CHARSET \
  # System database.
  SYSTEM_DB_HOST=$SYSTEM_DB_HOST \
  SYSTEM_DB_USER=$SYSTEM_DB_USER \
  SYSTEM_DB_PASSWORD=$SYSTEM_DB_PASSWORD \
  SYSTEM_DB_NAME=$SYSTEM_DB_NAME \
  SYSTEM_DB_CHARSET=$SYSTEM_DB_CHARSET

USER root
RUN apk update && \
  apk upgrade && \
  apk add git

RUN apk add --no-cache bash

# Change working dir to the server package.
WORKDIR /app/packages/server

RUN git clone https://github.com/vishnubob/wait-for-it.git

# Once we listen the mysql port run the migration task.
CMD ./wait-for-it/wait-for-it.sh $DB_HOST:3306 -- sh -c "node ./build/commands.js system:migrate:latest && node ./build/commands.js tenants:migrate:latest"
```

Ahson-Shaikh commented:
This is Mariadb Custom Dockerfile

```
FROM mariadb:10.2

USER root
ADD my.cnf /etc/mysql/conf.d/my.cnf

ARG MYSQL_DATABASE=default_database
ARG MYSQL_USER=default_user
ARG MYSQL_PASSWORD=secret
ARG MYSQL_ROOT_PASSWORD=root

ENV MYSQL_DATABASE=$MYSQL_DATABASE
ENV MYSQL_USER=$MYSQL_USER
ENV MYSQL_PASSWORD=$MYSQL_PASSWORD
ENV MYSQL_ROOT_PASSWORD=$MYSQL_ROOT_PASSWORD

# Copy init sql file with env vars and then the script will substitute the variables.
COPY ./init.sql /scripts/init.template.sql
COPY ./docker-entrypoint.sh /docker-entrypoint-initdb.d/docker-initialize.sh

# The scripts in the `docker-entrypoint-initdb.d/` directory are executed as
# the mysql user inside the MySQL Docker container.
RUN chown -R mysql:root /docker-entrypoint-initdb.d
RUN chown -R mysql:root /scripts

CMD ["mysqld"]
```

deiucanta commented:
@Ahson-Shaikh What's custom in each? what did you change? And why it was not working with the default images?

Ahson-Shaikh commented:
@deiucanta 

1. I made the nginx image bursting envs into that so EP can add Envs via the Environment. Previous envoyproxy was being used and not able to use Envs. 
2. The Mariadb mounts its own files for custom DB settings, with official ones, it wasn't connecting. With this, it connected. 
3. The migration service in the CMD section was hitting DB host like this, I changed it to Env, so the service starts for the migration.
Example for 3rd point: 
` CMD ./wait-for-it/wait-for-it.sh mysql:3306 -- sh -c "node ./build/commands.js system:migrate:latest && node ./build/commands.js tenants:migrate:latest"`
 
 I changed to 
 
`  CMD ./wait-for-it/wait-for-it.sh $DB_HOST:3306 -- sh -c "node ./build/commands.js system:migrate:latest && node ./build/commands.js tenants:migrate:latest"`

Ahson-Shaikh commented:
@deiucanta Additionally, the images I have changed were already made for Local only, I couldn't any image in Public repos.
